pyproject.toml

```
@@ -0,0 +1,59 @@
+[build-system]
+requires = ["setuptools>=64"]
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "pybop"
+version = "23.12"
+authors = [
+  {name = "The PyBOP Team"},
+]
+maintainers = [
+  {name = "The PyBOP Team"},
+]
+description = "Python Battery Optimisation and Parameterisation"
+readme = "README.md"
+license = { file = "LICENSE" }
+# https://pypi.org/classifiers/
+classifiers = []
+requires-python = ">=3.8, <=3.12"
+dependencies = [
+  "pybamm>=23.5",
+  "numpy>=1.16",
+  "scipy>=1.3",
+  "pandas>=1.0",
+  "pints>=0.5",
+]
+
+[project.optional-dependencies]
+plot = ["plotly>=5.0"]
+docs = [
+    "sphinx>=6",
+    "pydata-sphinx-theme",
+    "sphinx-autobuild",
+    "sphinx-autoapi",
+    "sphinx_copybutton",
+    "sphinx_favicon",
+    "sphinx_design",
+    "myst-parser",
+]
+all = ["pybop[plot]"]
```

```suggestion
all = ["pybop[plot, docs]"]
```

Looks like we missed adding the `docs` optional installation to `all`, let's update it here

---

Actually, we want to keep things like `[docs]` and `[dev]` or `[test]` dependencies separate from the `[all]` set. (PyBOP doesn't currently have a `[test]` extra, should I create one and add `pytest`, `pytest-mock`, and others there?)

I am unable to recall where this was referenced in the Scientific Python development guides, but from a rule-of-thumb this recommendation comes from the fact that more optional dependencies included all at once can potentially cause resolution conflicts with the core ones, plus `[docs]` dependencies increase the size of the dependency tree and subsequent interactions with `pip` will not be speedier (though the new tool `uv` counteracts all the latter by just being crazy fast). These extras are not used for the core package as a part of the API either, so they should be placed in isolated environments while testing (including them will break the isolation).

---

Good to know, I wasn't aware of that - let's keep it seperate then!

Taking a second look at the PyBaMM optional dependancies, I think implementing a similar `dev` extra that includes `pytest`, `pytest-mock`, `nox`, `pre-commit` and any others I missed would be a great way forward. 

Aside: I've also been looking at `uv`, adding an optional installation method in the `README` with it should be on our roadmap.

---

Thanks for your input! I'll make these changes soon.

`uv` is great with the experiments I've tried locally, we've opened an "up-for-grabs" issue for it over at PyBaMM. See https://github.com/pybamm-team/PyBaMM/issues/3825

---

Addressed in https://github.com/pybop-team/PyBOP/pull/203/commits/a6e32a16bf5f62c5f98c0909a872ca1fe395bbb4.